### DISCUSSION TITLE
Image Extraction...

### DISCUSSION THREAD

**Discussion by OlivierGruwe** [yan74/afplib#60] [Q&A]
Hi,

Great job !!

I'm surely missing something... How can i extract  images from afp using your lib ? imageData is always null...

IPD  number:12,offset:194,id:13889275,length:8201,rawData:null,charset:null,IOCAdat:[B@6f8e8894,imageData:null,

I'm using this piece of code :

   try (AfpInputStream in = new AfpInputStream(
                new BufferedInputStream(new FileInputStream(args[0])))) {

            SF sf;
            while ((sf = in.readStructuredField()) != null) {
                
                if(sf instanceof IPD) {
                    IPD ipd = (IPD) sf;
                    byte[] data = ipd.getImageData();

                    if(data != null)
                        System.out.println(data.length);
                }
                
                System.out.print(sf.eClass().getName() + "  ");
                printAttributes(sf);
                System.out.println();
                printContents(sf, 2);
            }

        } catch (IOException e) {
            e.printStackTrace();
        }

Could you help me ?

Thanks a lot

Olivier

**Comment by yan74** [comment]
Hi @OlivierGruwe,

you would have to use
```java
Triplet[] org.afplib.io.SDFHelper.ipds2sdf(List<IPD> ipds)
```
to get all SDFs contained in a sequence of IPDs (pass all IPDs you find in a row into that method). This gives you all the self defining fields (SDF) that make up the image. Now getting "image data" as in .jpg or .png that is a total different story and totally depends on how that image is build (tiled vs. non-tiled, b/w vs. banded image, does it use an image mask, ...). Please find details on that in the IOCA spec ([https://www.afpconsortium.org/uploads/1/1/8/4/118458708/ioca-reference-07.pdf](https://www.afpconsortium.org/uploads/1/1/8/4/118458708/ioca-reference-07.pdf)).
